### ISSUE TITLE
Add collection .IfAny() .IfSingle() .IfNotSingle() .IfNone() .IfAll()

### ISSUE THREAD

**Issue by mantinband** [mantinband/throw#20]
Add the following collection validations:

```csharp
var collection = new[] { 1, 2, 3, 3 };

collection.Throw().IfAny(item => item == 3);
collection.Throw().IfSingle(item => item == 2);
collection.Throw().IfNotSingle(item => item == 3);
collection.Throw().IfNone(item => item == 0);
collection.Throw().IfAll(item => item > 0);
```

**Comment by dwakel** [comment]
@mantinband 
Just to clarify if single is when only one element in the collection is equal to 2
And not single is when more than one element is equal to 3

**Comment by dwakel** [comment]
I picked up this issue. Currently writing test.
This is probably my first open-source contribution ever 
I've enjoyed working on this

**Comment by mantinband** [comment]
Awesome, that was quick! Let me know If you have any questions. Also, feel free to split this into multiple PRs if that would be more convenient 🙂

**Comment by dwakel** [comment]
@mantinband 
should these extensions be able to evaluate any type.

Im assuming for
collection.Throw().IfAny(item => item == 3);
It should accept an expression which can evaluate 
item => item > 3
item => item < 3
item => item <= 3
(Any inequality expression)

This however wont yield same results for a collection of strings unless a numeric property or function of the string (Count, Length, etc.) is Invoked

**Comment by mantinband** [comment]
I was thinking of something similar to this draft PR: https://github.com/mantinband/throw/pull/25
WDYT?

There is a C# language limitation that we cannot specify the method declaration as following:
```csharp
    public static ref readonly Validatable<TValue> IfAny<TValue, TItem>(this in Validatable<TValue<TItem>> validatable, Func<TItem, bool> predicate, [CallerArgumentExpression("predicate")] string? predicateName = null)
```

Unfortunately in this implementation, we need to disambiguate the predicate's input parameter (`int x` in the draft PR). Couldn't find a way around this..

**Comment by mantinband** [comment]
Perhaps this feature needs some more design thought, I can give it some thought next week. I'm sure there is some solution we can find to work around this (other than many extension methods for different collection types).
Feel free to take a different issue in the meantime if you want🙂

**Comment by mantinband** [comment]
@dwakel 
Take a look at the latest iteration of #25. This works as you would expect. The only downfall is that it boxes the Validatable struct which will hurt performance. I wonder if there is a way around this

**Comment by dwakel** [comment]
I noticed the validatable doesn't return TCollectionType but just encapsulates an IEnumerable directly.
It also didn't occur to me that you could just pass the inequality condition as a Func<item,bool> 😄
The picture I had in my mind was to accept an Expression type then evaluate it using LINQ

**Comment by dwakel** [comment]
Should these operations be limited to a subset ⊆ of Rational numbers

**Comment by mantinband** [comment]
Nah, I think the main use case will be on complex objects and not on collections of rational numbers/strings

**Comment by dwakel** [comment]
I'm going to jump back to working on this issue

**Comment by SakuraIsayeki** [comment]
Hi there! Just jumping into the pool. Is the issue still being worked on? Or is it up for grabs? 
I'm eager to contribute to this lib, so if there's anything I can add... Might as well start here ^^

**Comment by dwakel** [comment]
Hey @SakuraIsayeki you could definitely look at this issue.
Please also share any insights you have. It'll be nice knowing how you would go about implementing this

**Comment by Mitchman215** [comment]
Any update on this? It would be really useful to have this feature.
